casper/src/main/scala/coop/rchain/casper/util/rholang/Checkpoint.scala

```diff
@@ -2,29 +2,67 @@ package coop.rchain.casper.util.rholang
 
 import com.google.protobuf.ByteString
 
-import coop.rchain.crypto.codec.Base16
+import coop.rchain.catscontrib.TaskContrib._
+import coop.rchain.models.Par
+import coop.rchain.rholang.interpreter.Runtime
+import coop.rchain.rholang.interpreter.storage.StoragePrinter
+import coop.rchain.rspace.Blake2b256Hash
 
-import java.nio.file.Path
+import monix.execution.Scheduler
 
-class Checkpoint(val hash: ByteString, val location: Path, val size: Long) {
-  val dbLocation: Path = location.resolve(Base16.encode(hash.toByteArray))
+import scala.concurrent.SyncVar
+import scala.util.{Failure, Success, Try}
 
-  def toTuplespace(name: String): Tuplespace = {
-    val tsLocation = location.resolve(name)
+//runtime is a SyncVar for thread-safety, as all checkpoints share the same "hot store"
+class Checkpoint private (val hash: ByteString, runtime: SyncVar[Runtime]) {
```

Comment:
Have you thought about splitting the `Runtime` out of `Checkpoint`, given that you will have many `Checkpoint`s but only one `Runtime`?

Reply:
Define `type StateHash = ByteString`.

Also once we extract Runtime out of Checkpoint, we should remove the `val checkpoints: AtomicSyncVar[Map[ByteString, Checkpoint]]` as it will just become a map from StateHash to StateHash container, which is not much of a map.

Reply:
Note that it is not a value of the class, but only part of the constructor and so kind of included through the closure over methods. But my reasoning for including it in the constructor is that I wanted the control of the `take` and `put` calls would be encapsulated in one place, as they are a potential source for nasty bugs (as @henrytill knows because I asked him about one before making the PR). As @KentShikama shows below, we can refactor to take the `SyncVar[Runtime]` out of the constructor and pass it to function calls instead, but I don't think that makes the code better.

I suppose ultimately my viewpoint is that `Checkpoint`s should be able to be transformed into other `Checkpoint`s, which requires a `Runtime`. And the reason I want that is because that transformation is the whole reason we want checkpoints to begin with. Although, I am open to having my mind changed if someone wants to give a counter argument.

Reply:
@birchmd: good points vis-a-vis encapsulation.  I think your current solution is reasonable.